detekt/toolchains.bzl

```diff
@@ -0,0 +1,43 @@
+"""
+Macros for defining toolchains.
+See https://docs.bazel.build/versions/master/skylark/deploying.html#registering-toolchains
+"""
+
+load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kotlin_repositories", "kt_register_toolchains")
+load("@rules_java//java:repositories.bzl", "remote_jdk11_repos")
+load("@rules_jvm_external//:defs.bzl", "maven_install")
+load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")
+
+def rules_detekt_toolchains(detekt_version = "1.2.2"):
+    """Invokes `rules_detekt` toolchains.
+
+
+    Declares toolchains that are dependencies of the `rules_detekt` workspace.
+    Users should call this macro in their `WORKSPACE` file.
+
+    Args:
+        detekt_version: "io.gitlab.arturbosch.detekt:detekt-cli" version used by the rule.
+    """
+
+    remote_jdk11_repos()
+
+    kotlin_repositories()
+    kt_register_toolchains()
+
+    rules_proto_dependencies()
+    rules_proto_toolchains()
+
+    maven_install(
```

Comment:
oh damn, what about sha256s of these?

Reply:
I guess that’s irony but anyways. [Seems like `rules_jvm_external` checks this for us](https://github.com/bazelbuild/rules_jvm_external#checksum-verification).

Reply:
I am absolutely not kidding here, Bazel is about reproducibility

rules_jvm_external supports pinning https://github.com/bazelbuild/rules_jvm_external#pinning-artifacts-and-integration-with-bazels-downloader

I think we can include our default one and allow user to pass theirs

Reply:
We used to provide sha256 for detekt download, now we're allowing arbitrary files to get in, it still will properly invalidate caches if needed but that's not the way I'd like to see these rules going

Reply:
Sure, I’ll take a look.

Reply:
The annoying thing with rules_jvm_external's pinning is that it works by generating a macro when you call `maven_install`... Macro that you have to call. Which you cannot do inside of a macro, nor before it's generated (i.e. you cannot do that inside of the same file, if it's not the `WORKSPACE`).

Reply:
@artem-zinnatullin, I’m shelving pinning for now. I’ve tried two solutions and both are weird.

* Pinning. Forces users to declare a third `load` and invoke a third macro (like `rules_detekt_pin` or something) — no matter they want to use pinning or not. Then, the Detekt version change. We can either force users to regenerate pinned JSON and pass it to us or allow them to ignore pinning. The regenerating approach will force them to include a weird JSON file in the VCS (putting aside the weird machinery behind the generation procedure). The ignoring approach is error-prone since in this case the pinning macro should not be invoked but in the default configuration it should...
* Nuking Kotlin. I’ve actually rewrote the wrapper in Java but it felt like I’ve exchanged the dependency for a lack of safety (`null`, general call flow). Plus I doubt that we’ll be able to survive without dependencies forever — for example, I wanted to use RxJava for multiplexing (see #41).

I’m open to revisit this later when the issue will be more evident. For now `rules_jvm_external` will check hashes in Maven repos we’ve provided (and those cannot be user-changed).

Reply:
Given that the dependency fetched through Maven is Detekt itself and not Kotlin, how would switching from Kotlin to Java help avoiding this dependency?

Reply:
@Kernald, I was referring to a potential workaround of downloading a fat Detekt JAR as a `http_file`, then using it as a `java_import`. Right now it’s dangerous cause this JAR bundles Kotlin runtime and the wrapper does the same. Switching to Java leaves us with zero dependencies and a safe way to include the fat JAR.

Reply:
Got it, thanks!